src/dss/create_notebook.py

```diff
@@ -0,0 +1,71 @@
+from pathlib import Path
+
+from charmed_kubeflow_chisme.kubernetes import KubernetesResourceHandler
+from lightkube import Client
+from lightkube.resources.apps_v1 import Deployment
+from lightkube.resources.core_v1 import Service
+
+from dss.config import (
+    DSS_CLI_MANAGER_LABELS,
+    DSS_NAMESPACE,
+    FIELD_MANAGER,
+    MANIFEST_TEMPLATES_LOCATION,
+    NOTEBOOK_PVC_NAME,
+)
+from dss.logger import setup_logger
+from dss.utils import get_mlflow_tracking_uri, get_notebook_url, wait_for_deployment_ready
+
+# Set up logger
+logger = setup_logger("logs/dss.log")
+
+
+def create_notebook(name: str, image: str, lightkube_client: Client) -> None:
+    """
+    Creates a Notebook server on the Kubernetes cluster.
+
+    Args:
+        name (str): The name of the notebook server.
+        image (str): The image used for the notebook server.
+        lightkube_client (Client): The Kubernetes client.
+    """
+    manifests_file = Path(
+        Path(__file__).parent, MANIFEST_TEMPLATES_LOCATION, "notebook_deployment.yaml.j2"
+    )
+
+    config = _get_notebook_config(image, name)
+
+    # Initialize KubernetesResourceHandler
+    k8s_resource_handler = KubernetesResourceHandler(
+        field_manager=FIELD_MANAGER,
+        labels=DSS_CLI_MANAGER_LABELS,
+        template_files=[manifests_file],
+        context=config,
+        resource_types={Deployment, Service},
+        lightkube_client=lightkube_client,
+    )
+
+    try:
+        k8s_resource_handler.apply()
+
+        wait_for_deployment_ready(lightkube_client, namespace=DSS_NAMESPACE, deployment_name=name)
+
+        logger.info(f"Notebook {name} created.")
+    except TimeoutError as err:
+        logger.error(f"{str(err)}\nCleaning up the resources just created.")
+        k8s_resource_handler.delete()
```

Comment:
I noticed that if the `create-notebook` command times out, it deletes everything in the `dss` namespace. The namespace is left intact, though. Could it be that we have the same labels everywhere and the `delete()` method catches it and just wipes out everything?

Reply:
(I'm not sure deleting anything is the right choice if there's a timeout as it makes debugging hard.  But ignoring that separate conversation for now and assuming we do want to delete things on a timeout...) 

This is a good catch, and turns out its broken more than just this way :D

Yes the issue here is that the `DSS_CLI_MANAGER_LABELS` labels are used for anything the manager did (both from `initialize` and `create-notebook`).  So the `krh.delete()` here deleted all the Deployment and Services from both `initialize` and `create-notebook`.  The namespace (and pvc) were not deleted because the notebook's KRH is scoped only to deployments and services (it will never delete objects of other types).  

Another thing that could have gone wrong and I missed was this can happen:
1) `dss create-notebook notebook1` 
    1) assume this succeeds and we now have `notebook1`
2) `dss create-notebook notebook2`
    1) assume this times out, triggering deletion

On timeout, (2.i) will delete `notebook2` **and** `notebook1` (...and mlflow)

Reply:
I think the question here is what functionality do we need? We should have an easy way to:
1. delete all dss resources
2. delete a single notebook and its associated resources (`Service`)

I think labels are pretty good for (1).  The remove commands for dss could look for our `DSS_CLI_MANAGER_LABELS` in any object type we might create.  

For (2), we should be using whatever #34 / #51 do.  Otherwise we're just creating a second system of removal.  So maybe in this PR we should just log that something timed out, and then add the deletion in #51.  wdyt?

Reply:
> So maybe in this PR we should just log that something timed out, and then add the deletion in https://github.com/canonical/data-science-stack/pull/51. wdyt?

I think it is fine if we log that something timed out, but then we'll hit the case where the resource is not deleted and then after a while it could be up and ready without noticing. I'd say let's not remove, log the timeout, but warn people that the notebook server might be in the cluster and ways to check it (`dss list`)? It feels awkward, but at least it gives users a bit more visibility.

Kind of related:
While I agree that having the `dss create-notebook` command to output the notebook server address is really handy, and for knowing that address we need a timeout, I am wondering if the timeout is a good thing to have. First of all, users do not have control on the timeout time (as mentioned in #44), and even if they did, there are things that don't entirely depend on time: lightkube errors, memory, disk space, even gpus, internet connectivity, etc.
I know this was not in the spec, but perhaps a better approach would be to have something similar to what microk8s or juju does with status to show the notebook servers by name, if they are running or not, and their address. This way we avoid having to delete deployments and services, as users will have full access to the state of notebook servers.

If you feel this is out of scope of this PR, I suggest we file an issue to follow up on this.

Reply:
The deletion on timeout is indeed a problem when there's an unexpected error, then we will not be able to debug it.
For this PR, I will move on with @DnPlas 's suggestion to log the timeout, but warn people that the notebook server might be in the cluster to check it with `dss list`.

Whether we should have timeout or not is a discussion that we should have as a team and get feedback from the UX team on. I'd say even update the spec about this. cc: @misohu 
Let me file an issue to track this and let's discuss further there.

Reply:
filed #63 to discuss the timeout